File: tools/verify_qr.py
"""Decode the QR-concealed-tag matrix: does the QR still scan, and is the tag still found?

Prints a pass/fail grid per QR version so you can pick a configuration before printing.
QR decoders differ a lot; where available, WeChat's is closer to what a phone does than
OpenCV's built-in one, so it is reported separately rather than averaged in.
"""
import json, os, sys
import numpy as np, cv2

os.chdir(os.path.join(os.path.dirname(__file__), '..'))
d = json.load(open('tools/verify_qr_dump.json'))
PAYLOAD, TAG_ID = d['payload'], d['tagId']

qr_det = cv2.QRCodeDetector()
tag_det = cv2.aruco.ArucoDetector(cv2.aruco.getPredefinedDictionary(cv2.aruco.DICT_APRILTAG_36h11))

wechat = None
try:
    wechat = cv2.wechat_qrcode.WeChatQRCode()
    print('WeChat QR detector: available')
except Exception as e:
    print(f'WeChat QR detector: unavailable ({type(e).__name__}) — using OpenCV only')

def decode_qr(img):
    try:
        txt, pts, _ = qr_det.detectAndDecode(img)
        if txt == PAYLOAD:
            return True
    except cv2.error:
        pass
    if wechat is not None:
        try:
            res, _ = wechat.detectAndDecode(img)
            if any(r == PAYLOAD for r in res):
                return True
        except cv2.error:
            pass
    return False

def find_tag(img):
    _, ids, _ = tag_det.detectMarkers(img)
    return ids is not None and TAG_ID in ids.flatten().tolist()

rows = []
for c in d['cases']:
    if 'error' in c:
        rows.append({**c, 'qr': False, 'tag': False})
        continue
    img = np.array(c['img'], np.uint8)
    # pad with white: detectors need clean margin beyond the symbol's own quiet zone
    img = np.pad(img, 40, constant_values=255)
    rows.append({**{k: v for k, v in c.items() if k != 'img'},
                 'qr': decode_qr(img), 'tag': find_tag(img)})

ratios = sorted({r['ratioPct'] for r in rows})
versions = sorted({r['requestedVersion'] for r in rows})
print(f'\npayload {PAYLOAD!r}, hidden tag {d["tagDict"]} #{TAG_ID}')
print('cell = QR / TAG   (o = ok, . = fail)\n')
for v in versions:
    sub = [r for r in rows if r['requestedVersion'] == v]
    n = sub[0].get('n', '?')
    ver = sub[0].get('version', '?')
    print(f'QR v{ver} ({n} modules){"  [auto]" if v == 0 else ""}')
    print('        ' + ''.join(f'{p:>8}%' for p in ratios))
    for ecc in ['L', 'M', 'Q', 'H']:
        cells = []
        for p in ratios:
            m = next((r for r in sub if r['ecc'] == ecc and r['ratioPct'] == p), None)
            if m is None:
                cells.append('     -  ')
            elif m.get('error'):          # payload too big for this version+ecc
                cells.append('   n/a  ')
            else:
                cells.append(f'{"o" if m["qr"] else ".":>5} /{"o" if m["tag"] else "."}')
        print(f'  {ecc}   ' + ''.join(cells))
    areas = {p: next((r['areaPct'] for r in sub if r['ratioPct'] == p), None) for p in ratios}
    print('  area% ' + ''.join(f'{areas[p]:>8}' for p in ratios) + '\n')

# ---- headline: largest overlay that keeps both working, per ecc, on the auto version ----
print('Largest tag size where QR still scans AND tag is still found (auto version):')
best = {}
for ecc in ['L', 'M', 'Q', 'H']:
    ok = [r['ratioPct'] for r in rows
          if r['requestedVersion'] == 0 and r['ecc'] == ecc and r['qr'] and r['tag']]
    best[ecc] = max(ok) if ok else None
    print(f'  {ecc}: ' + (f'{best[ecc]}% of QR width' if best[ecc] else 'none of the tested sizes worked'))

# ---- nested stacks ----
print('\nNested stacks (each layer carries the next in its middle):')
os.makedirs('tools/nested_out', exist_ok=True)
for c in d.get('nested', []):
    if c.get('error'):
        print(f'  {c["name"]:26} compose failed — {c["error"]}')
        continue
    img = np.pad(np.array(c['img'], np.uint8), 60, constant_values=255)
    qr_ok = decode_qr(img) if c['hasQr'] else None
    tag_ok = find_tag(img)
    safe = c['name'].replace(' ', '').replace('>', '-')
    cv2.imwrite(f'tools/nested_out/{safe}.png', img)
    qr_txt = 'n/a' if qr_ok is None else ('ok' if qr_ok else 'FAIL')
    print(f'  {c["name"]:26} depth {c["depth"]}  QR {qr_txt:>4}   tag {"ok" if tag_ok else "FAIL"}')
    for n in c['notes']:
        print(f'{"":30}- {n}')

json.dump(rows, open('tools/verify_qr_results.json', 'w'), indent=1)
fails = [r for r in rows if r.get('error')]
print(f'\n{len(rows)} configurations tested, {len(fails)} failed to compose')
sys.exit(0)

File: tools/verify_qr.mjs
// Compose a matrix of QR-concealed tags and dump them as pixel grids for OpenCV to
// decode. Sweeps ECC level x overlay ratio x QR version.
import { readFileSync, writeFileSync } from 'fs';
import { fileURLToPath } from 'url';
import path from 'path';
import vm from 'vm';

const root = path.dirname(path.dirname(fileURLToPath(import.meta.url)));
const ctx = { window: {}, console, Math };
vm.createContext(ctx);
for (const f of ['vendor/qrcode-generator.min.js', 'data/aruco_dicts.js', 'data/apriltag3.js',
                 'js/runetag.js', 'js/render.js', 'js/qrapril.js', 'js/nest.js']) {
  vm.runInContext(readFileSync(path.join(root, f), 'utf8'), ctx, { filename: f });
}
const Render = vm.runInContext('Render', ctx);
const QrApril = vm.runInContext('QrApril', ctx);

const PAYLOAD = 'https://klyrai.github.io/tagforge/';
const TAG_DICT = 'april_36h11', TAG_ID = 23;
const PX = 8;   // pixels per QR module in the dump

function raster(tile, pxPerUnit) {
  const size = Math.round(tile.wu * pxPerUnit);
  const img = Array.from({ length: size }, () => new Array(size).fill(255));
  for (const s of tile.shapes) {
    const v = s.black ? 0 : 255;
    if (s.t === 'rect') {
      const x0 = Math.round(s.x * pxPerUnit), y0 = Math.round(s.y * pxPerUnit);
      const x1 = Math.round((s.x + s.w) * pxPerUnit), y1 = Math.round((s.y + s.h) * pxPerUnit);
      for (let y = Math.max(0, y0); y < Math.min(size, y1); y++)
        for (let x = Math.max(0, x0); x < Math.min(size, x1); x++) img[y][x] = v;
    } else {
      for (let y = 0; y < size; y++)
        for (let x = 0; x < size; x++) {
          const dx = x + 0.5 - s.cx * pxPerUnit, dy = y + 0.5 - s.cy * pxPerUnit;
          if (dx * dx + dy * dy <= (s.r * pxPerUnit) ** 2) img[y][x] = v;
        }
    }
  }
  return img;
}

const cases = [];
const inner = Render.gridTag(TAG_DICT, TAG_ID, 1);
for (const version of [0, 3, 5, 8]) {
  for (const ecc of ['L', 'M', 'Q', 'H']) {
    for (const ratioPct of [20, 25, 30, 35, 40, 45, 50]) {
      let tile;
      try {
        tile = QrApril.compose({ payload: PAYLOAD, version, ecc, innerTile: inner,
                                 innerLabel: `${TAG_DICT} #${TAG_ID}`, ratioPct, quiet: 4 });
      } catch (e) {
        // qrcode-generator throws bare strings ("code length overflow") for
        // payloads that exceed the chosen version's capacity.
        cases.push({ requestedVersion: version, ecc, ratioPct, error: String(e && e.message || e) });
        continue;
      }
      cases.push({
        version: tile.qr.version, requestedVersion: version, ecc, ratioPct,
        n: tile.qr.n, overlayModules: tile.qr.overlayModules,
        areaPct: +tile.qr.areaPct.toFixed(1), budgetPct: tile.qr.budgetPct,
        img: raster(tile, PX),
      });
    }
  }
}

// ---- nested stacks: each layer carries the next in its middle ----
const Nest = vm.runInContext('Nest', ctx);
const qrL = (r, ecc = 'H') => ({ type: 'qr', payload: PAYLOAD, version: 0, ecc, ratioPct: r, quiet: 4 });
const runeL = () => ({ type: 'runetag', ordinal: 1, quietPct: 5 });
const leaf = { kind: 'grid', dict: TAG_DICT, id: TAG_ID };

// Nested tiles differ hugely in abstract width, so rasterize each to a fixed pixel size
// — otherwise a RuneTag outer layer lands at a fraction of the resolution of a QR outer.
const NEST_PX = 900;
const stacks = {
  'qr(30) > tag':      [qrL(30)],
  'qr(40) > tag':      [qrL(40)],
  'qr > tag':          [qrL(35)],
  'runetag > tag':     [runeL()],
  'runetag > qr > tag': [runeL(), qrL(35)],
  'runetag > qr(45) > tag': [runeL(), qrL(45)],
  'qr(50) > runetag > tag': [qrL(50), runeL()],
  'qr(40) > runetag > tag': [qrL(40), runeL()],
};
const nested = [];
for (const [name, stack] of Object.entries(stacks)) {
  try {
    const tile = Nest.compose(stack, leaf);
    nested.push({ name, label: tile.stack.label, notes: tile.stack.notes, depth: tile.stack.depth,
                  hasQr: stack.some(l => l.type === 'qr'),
                  img: raster(tile, NEST_PX / tile.wu) });
  } catch (e) {
    nested.push({ name, error: String(e && e.message || e) });
  }
}

writeFileSync(path.join(root, 'tools', 'verify_qr_dump.json'),
  JSON.stringify({ payload: PAYLOAD, tagDict: TAG_DICT, tagId: TAG_ID, px: PX, cases, nested }));
console.log(`dumped ${cases.length} matrix configurations, ${nested.length} nested stacks`);
